CMakeLists.txt

```diff
@@ -70,6 +70,14 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
   set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Choose the type of build." FORCE)
 endif()
 
+
+if(CMAKE_BUILD_TYPE MATCHES "Release" AND NOT MSVC)
+  set(CMAKE_CXXFLAGS "-g -O3 -march-native -ffast-math -Wfatal-errors")
+elseif(NOT MSVC)
+  set(CMAKE_CXX_FLAGS_DEBUG "-Wall -g -O2 -Wfatal-errors -fsanitize=address -fsanitize=undefined -fno-omit-frame-pointer")
+endif()
+
```

Comment:
I don't think it's a good idea to hardcode compiler flags that are gcc/clang specific. adios2 may be built with xlc, intel, cray, pgi, etc.

I do think the sanitizers are useful for testing / the CI, but shouldn't necessarily be in every debug build. And I prefer `-O0` in my debug builds.

The rest of this patch looks good to me.

Reply:
Yup, you're exactly right about compiler detection; that's fixed.

As to `-O2`: I wondered how y'all felt about the how long the unit tests took; sometimes `-O0` makes trillions of superfluous stack writes that are pretty frustrating unless you're under a debugger. But if everyone is ok with `-O0` I'm ok with it too.

Reply:
I think @chuckatkins should take a look at this. I have no say, anyway, I'm just a contributor. I think even after the change above, you're still setting CMAKE_CXX_FLAGS irregardless of clang/gcc in the Release case.

But I'm also generally not sure about setting CMAKE_CXX_FLAGS in CMakeLists.txt in the first place -- does that mean if I specify those on the command line, they'll be overwritten regardless, or does the command line take precedence?

Reply:
This entire block needs to be taken out.  The way you target CPU sets is different on different platforms and the ADIOS project is not in the business of encoding compiler secret sauce.  For example, this is handled in an entirely different mechanism when building on a cray.  CMake has default flags for debug and release across a wide swatch of compilers that's been heavily tested for some time.  While it may not the "super optimal" it certainly gives a safe best guess for optimized configurations; i.e. O3 isn't universally safe, the fastmath and  options aren't available on all the compilers we are supporting, you've dropped the -DNDEBUG flag often used by std library implementations, etc.  If a user wants to specify some specific special sauce flags when they build it then they can certainly do so via the `CFLAGS`, `CXXFLAGS`, and `FFLAGS` environment variables or `-DCMAKE_CXX_FLAGS`, `-DCMAKE_C_FLAGS`, and `-DCMAKE_Fortran_FLAGS` CMake variables (yes, we use all three languages which, while not the point I'm making, would need to be accounted for if this were to go in, which I'm not supportive of).  We do just that in some of our dashboard builds but it's compiler and environment specific and isn't suitable to encode directly into adios.

Reply:
This entire block needs to be taken out. The way you target CPU sets is different on different platforms and the ADIOS project is not in the business of encoding compiler secret sauce. For example, this is handled in an entirely different mechanism when building on a cray. CMake has builtin default flags for debug and release across a wide swath of compilers that's been heavily tested for some time. While it may not the "super optimal" it certainly gives a safe best guess for optimized configurations; i.e. O3 isn't universally safe, the fastmath and W options aren't available on all the compilers we are supporting, you've dropped the -DNDEBUG flag often used by std library implementations, etc. If a user wants to specify some specific special sauce flags when they build it then they can certainly do so via the CFLAGS, CXXFLAGS, and FFLAGS environment variables or -DCMAKE_CXX_FLAGS, -DCMAKE_C_FLAGS, and -DCMAKE_Fortran_FLAGS CMake variables (yes, we use all three languages which, while not the point I'm making, would need to be accounted for if this were to go in, which I'm not supportive of). We do just that in some of our dashboard builds but it's compiler and environment specific and isn't suitable to encode directly into adios.